lib/src/builder/builder.dart

```diff
@@ -1,8 +1,144 @@
 import 'package:build/build.dart';
-import 'package:source_gen/source_gen.dart';
+import 'package:dart_style/dart_style.dart';
+import 'package:over_react/src/builder/generation/declaration_parsing.dart';
+import 'package:over_react/src/builder/generation/impl_generation.dart';
+import 'package:path/path.dart' as p;
+import 'package:source_span/source_span.dart';
 
 import './util.dart';
-import './generator.dart';
 
-Builder overReactBuilder(BuilderOptions options) =>
-    new PartBuilder([new OverReactGenerator()], outputExtension);
+Builder overReactBuilder(BuilderOptions options) => new OverReactBuilder();
+
+class OverReactBuilder extends Builder {
+  static final _headerLine = '// '.padRight(77, '*');
+
+  static final _formatter = new DartFormatter();
+
+  static final RegExp _overReactPartDirective = new RegExp(
+    r'''['"].*\.over_react\.g\.dart['"]''',
+  );
+
+  static bool _mightContainDeclarations(String source) {
+    return ParsedDeclarations.key_any_annotation.hasMatch(source) ||
+      _overReactPartDirective.hasMatch(source);
+  }
+
+  @override
+  FutureOr<void> build(BuildStep buildStep) async {
+    final source = await buildStep.readAsString(buildStep.inputId);
+    if (!_mightContainDeclarations(source)) {
+      return;
+    }
+
+    CompilationUnit libraryUnit;
+    try {
+      libraryUnit = parseCompilationUnit(
+        source,
+        name: buildStep.inputId.path,
+        suppressErrors: false,
+        parseFunctionBodies: true);
+    } catch (error, stackTrace) {
+      log.severe('There was an error parsing the compilation unit for file: ${buildStep.inputId}');
+      log.fine(error);
+      log.fine(stackTrace);
+      return;
+    }
+
+    if (isPart(libraryUnit)) {
+      return;
+    }
+
+    final librarySourceFile = new SourceFile.fromString(
+      source,
+      url: idToPackageUri(buildStep.inputId));
+    final libraryDeclarations = new ParsedDeclarations(
+      libraryUnit,
+      librarySourceFile,
+      log);
+    if (libraryDeclarations.hasErrors) {
+      log.severe('There was an error parsing the file declarations for file: ${buildStep.inputId}');
+      return;
+    }
+
+    final outputs = <String>[];
+    final libraryGenerator = new ImplGenerator(log, librarySourceFile)
+      ..generate(libraryDeclarations);
+    outputs.add(libraryGenerator.outputContentsBuffer.toString());
+
+    final parts = libraryUnit.directives
+      .whereType<PartDirective>()
+      // Ignore `.over_react.g.dart` parts - that's what we're generating.
+      .where((part) => !part.uri.stringValue.endsWith(outputExtension));
+
+    for (final part in parts) {
+      final partId = new AssetId.resolve(
+        part.uri.stringValue,
+        from: buildStep.inputId);
+      final partSource = await buildStep.readAsString(partId);
+      final partSourceFile = new SourceFile.fromString(
+        partSource,
+        url: idToPackageUri(partId));
+      CompilationUnit partUnit;
+      try {
```

Comment:
This code is very similar to the parsing and generating for the main library code above. What's the reason not to DRY this up?

Reply:
Yeah, the main difference is that I have to parse the compilation unit for the library earlier so that I can check the directives for a PartOfDirective (i.e. to see if the file is actually a part file) so that we can bail early. I can probably dry it up though